docs/internal_abi.md

```diff
@@ -0,0 +1,323 @@
+
+In C, `struct` layout is a trade between storage density and accidental aliasing.
+If two fields are accessed concurrently, false sharing in the cache can cause significant performance bottlenecks.
+In Verona; however, no mutable objects are ever shared and so, aside from reference count manipulations on immutable objects and interactions between adjacently allocated objects in different regions, there are no negative interactions with modern caches.
+The Verona field layout algorithm is therefore free to optimise for space.
+
+All fields whose size is a power-of-two should be sorted by size and laid out largest to smallest.
+The end of each section is more strongly aligned than the next requires and so this does not require any padding.
+
+Next, any union types that have a power-of-two sized payload but a smaller discriminator should be split into payload and discriminator.
+The payloads should be inserted into the sections for the same sized power-of-two objects.
+The discriminators should be packed into a single larger value.
+
+*Rationale:* Extracting from a bitfield is no more expensive if the bitfield is densely packed and if a function is pattern matching on multiple fields then they may be able to reuse a register for the descriptor values.
+
+The remaining values should be stored in the order in which they are declared, between the power-of-two region and the descriptors.
+
+*Rationale:* We could do some clever bin packing here but systems programmers who are using odd-sized storage-only types are likely to be careful about arranging them usefully already.
+A future version can improve bin packing and not break anything.
+
+
+*Note:* Java implementations typically pack all pointers together such that the trace logic just needs to look at a range, rather than have data or code describing the complete layout of the object.
+This may be possible for Verona, depending on the constraints for value types as fields and on pointers to value types.
+If value types cannot be passed by reference then they do not need to be stored contiguously in fields, because they will always be either copied when passed down the stack or accessed from functions that are aware of the location of their components.
+
+External references
+-------------------
+
+External references are structures with the same header as a Verona object followed by two pointers, one to the table that stores the external reference and one to the object.
+The rules for operations on this are defined in [`externalreference.h`](../src/rt/region/externalreference.h).
+External references are never embedded in other structures, they are always referenced by pointer and so fields or arrays of external references follow the same rules as any other kind of object pointer.
+
+Arrays
+------
+
+Arrays are objects (with the standard header) followed by a run of values of the same type.
+
+*Open question*: Should `Array[Bool]` be equivalent to `Array[U1]` or `Array[U8]` in storage?
```

Comment:
Similar to `std::vector<bool>`, I think this could be an optimisation. First we implement as `U8`, then we optimise if necessary.

Reply:
Agreed for the initial version, it remains an open question longer term.